### PR TITLE
Allow TORCH_LOGS rank filtering via TORCH_LOGS_RANKS

### PR DESCRIPTION
**Pull request by xmfan**
Addresses issue #117383 

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #117688

e.g. TORCH_LOGS_RANKS=0 TORCH_LOGS="+dynamo" torchrun --nproc_per_node=8 t.py
```
[rank0]:[2024-01-17 13:55:29,864] [0/0] torch._dynamo.symbolic_convert: [DEBUG] INLINING <code object _call_impl at 0x7fae3d8e1840, file "/home/xmfan/core/pytorch/torch/nn/modules/module.py", line 1513>, inlined according skipfiles.MOD_INLINELIST                                                                             
[rank0]:[2024-01-17 13:55:29,866] [0/0] torch._dynamo.symbolic_convert.__trace_source: [DEBUG] TRACE starts_line /home/xmfan/core/pytorch/torch/nn/modules/module.py:1513 in _call_impl (Module._call_impl) (inline depth: 1)                                                                                                      
[rank0]:[2024-01-17 13:55:29,866] [0/0] torch._dynamo.symbolic_convert.__trace_source: [DEBUG]         def _call_impl(self, *args, **kwargs):                                                                                                                                                                                      
[rank0]:[2024-01-17 13:55:29,888] [0/0] torch._dynamo.symbolic_convert.__trace_source: [DEBUG] TRACE starts_line /home/xmfan/core/pytorch/torch/nn/modules/module.py:1514 in _call_impl (Module._call_impl) (inline depth: 1)                                                                                                      
[rank0]:[2024-01-17 13:55:29,888] [0/0] torch._dynamo.symbolic_convert.__trace_source: [DEBUG]             forward_call = (self._slow_forward if torch._C._get_tracing_state() else self.forward)                                                                                                                                  
[rank0]:[2024-01-17 13:55:29,888] [0/0] torch._dynamo.symbolic_convert: [DEBUG] TRACE LOAD_GLOBAL torch []                                                                                                                                                                                                                         
[rank0]:[2024-01-17 13:55:29,889] [0/0] torch._dynamo.symbolic_convert: [DEBUG] TRACE LOAD_ATTR _C [PythonModuleVariable(<module 'torch' from '/home/xmfan/core/pytorch/torch/__init__.py'>)]                                                                                                                                      
[rank0]:[2024-01-17 13:55:29,889] [0/0] torch._dynamo.symbolic_convert: [DEBUG] TRACE LOAD_ATTR _get_tracing_state [PythonModuleVariable(<module 'torch._C' from '/home/xmfan/core/pytorch/torch/_C.cpython-310-x86_64-linux-gnu.so'>)]                                                                                            
[rank0]:[2024-01-17 13:55:29,889] [0/0] torch._dynamo.symbolic_convert: [DEBUG] TRACE CALL_FUNCTION 0 [TorchInGraphFunctionVariable(<built-in method _get_tracing_state of PyCapsule object at 0x7fae40b53f30>)]                                                                                                                   
[rank0]:[2024-01-17 13:55:29,889] [0/0] torch._dynamo.symbolic_convert: [DEBUG] TRACE POP_JUMP_IF_FALSE 16 [ConstantVariable(NoneType)]                                                                                                                                                                                            
[rank0]:[2024-01-17 13:55:29,889] [0/0] torch._dynamo.symbolic_convert: [DEBUG] TRACE LOAD_FAST self []                                                                                                                                                                                                                            
[rank0]:[2024-01-17 13:55:29,889] [0/0] torch._dynamo.symbolic_convert: [DEBUG] TRACE LOAD_ATTR forward [UnspecializedNNModuleVariable(DistributedDataParallel)]                                                                                                                                                                   
[rank0]:[2024-01-17 13:55:29,890] [0/0] torch._dynamo.symbolic_convert: [DEBUG] TRACE STORE_FAST forward_call [UserMethodVariable(<function DistributedDataParallel.forward at 0x7fae3d231240>, UnspecializedNNModuleVariable(DistributedDataParallel))]                                                                           
[rank0]:[2024-01-17 13:55:29,890] [0/0] torch._dynamo.symbolic_convert.__trace_source: [DEBUG] TRACE starts_line /home/xmfan/core/pytorch/torch/nn/modules/module.py:1517 in _call_impl (Module._call_impl) (inline depth: 1)
```

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

### CODE REVIEW

File: torch/_logging/_internal.py [hunk 1]
```diff
@@ -738,9 +741,40 @@ def _default_formatter():
...
+
+        return True
+
+
+def _default_filter() -> Optional[TorchLogsFilter]:
+    ranks_str = os.environ.get(LOG_RANK_ENV_VAR, None)
+    if ranks_str is None:
+        return None
+
+    ranks = [int(rank_str) for rank_str in ranks_str.split(",")]
```
Comment (wconstab) [inline torch/_logging/_internal.py:761]: maybe `try:` around  this parsing bit, and then in the 'except' clause you can raise the ValueError you have?

**Comment by wconstab** [comment]
lgtm but lets circulate more widely and get feedback

File: torch/_logging/_internal.py [hunk 2]
```diff
@@ -738,9 +741,40 @@ def _default_formatter():
...
+    return TorchLogsFilter(ranks)
+
+
+DEFAULT_FILTER = _default_filter()
+
+
 def _setup_handlers(create_handler_fn, log):
     debug_handler = _track_handler(create_handler_fn())
     debug_handler.setFormatter(DEFAULT_FORMATTER)
+    if DEFAULT_FILTER:
```
Comment (wanchaol) [inline torch/_logging/_internal.py:776]: don't we always go into this if branch as `DEFAULT_FILTER` is a global variable? maybe we can remove the if clause if that's the case.

**Review by wanchaol — APPROVED** [review]
LGTM, imo this is an awesome feature that would solve a lot of our pains when using torch.compile + distributed together!

File: torch/_logging/_internal.py (hunk 2 above)
Comment (xmfan) [reply torch/_logging/_internal.py:776]: I think there's some additional overhead to run each log through a no-op filter, so I only attach the filter when it is needed

**Comment by wconstab** [comment]
i think we gave sufficient chance for API feedback. lets land it and also pls do a post about it so people try it out and give more feedback

**Comment by bdhirsh** [comment]
I am looking forward to this PR!

**Comment by xmfan** [comment]
I opted to instead implement this change in torchrun: https://github.com/pytorch/pytorch/pull/118562. Implementing it in the process launcher additionally allows for filtering of non-TORCH_LOGS logs such as c++ or logs outside of pytorch. Given that the torchrun implementation is a superset of this one, and everyone I've spoken to seems to use torchrun or call its underlying api torch.distributed.run directly, I'll close this PR shortly.

**Pull request closed without merging**